repair/repair.cc

```diff
@@ -1959,23 +1985,66 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_live_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    size_t nr_racks = source_dc_racks.size();
+                    size_t max_replicas_per_rack = (rf + nr_racks - 1) / nr_racks;
+                    size_t possibly_lost_replicas = 0;
+                    for (const auto& [rack, nr] : lost_nodes_per_rack) {
+                        possibly_lost_replicas += std::min(nr, max_replicas_per_rack);
+                    }
+                    // Can we ensure a quorum in source_dc?
+                    if (possibly_lost_replicas < rf / 2) {
+                        ks_source_dc = source_dc;
+                    } else if (reason == streaming::stream_reason::rebuild) {
+                        if (rf == 0) {
+                            throw std::runtime_error(fmt::format("Cannot rebuild with source_dc={}: Replication factor in this data center is 0."
+                                    " Use a different source_dc, or no source_dc option to consider the whole cluster for repair", source_dc));
+                        }
+                        rlogger.warn("{}: keyspace={}: Cannot ensure quorum in source_dc={}: possibly lost {} replicas out of {}: rebuilding using source_dc anyway",
+                                op, keyspace_name, source_dc, possibly_lost_replicas, rf);
+                        ks_source_dc = source_dc;
+                    } else {
+                        rlogger.warn("{}: keyspace={}: Cannot ensure quorum in source_dc={}: possibly lost {} replicas out of {}: Falling back to full-cluster repair",
+                                op, keyspace_name, source_dc, possibly_lost_replicas, rf);
+                    }
+                    break;
+                }
+                case locator::replication_strategy_type::everywhere_topology:
+                    ks_source_dc = source_dc;
+                    break;
+                case locator::replication_strategy_type::simple:
+                    // With simple strategy, we have no assurance that source_dc will contain
+                    // another replica for all token ranges.
+                    break;
+                default:
+                    break;
+                }
+            } else if (reason == streaming::stream_reason::rebuild && !source_dc.empty()) {
+                throw std::runtime_error(fmt::format("Cannot rebuild with source_dc={}: No live nodes remained in the data center."
+                        " Use a different source_dc or no source_dc option to consider the whole cluster for repair", source_dc));
```

Comment:
This is not accurate. When user provides no source_dc option, Scylla will choose anyone dc. By default, scylla chooses the local dc. 


```
        <src-dc-name>
            Name of DC from which to select sources for streaming. By default,
            pick any DC
```

Reply:
I see.  That's in :
```
future<> repair_service::rebuild_with_repair(locator::token_metadata_ptr tmptr, sstring source_dc) {
    ...
    if (source_dc.empty()) {
        auto& topology = tmptr->get_topology();
        source_dc = topology.get_datacenter();
    }
```

So we should probably fall back to all-cluster too if no source_dc was given and the local dc cannot guarantee quorum, like we do in this series for replace.

Reply:
See v5.
Added a more qualified source_dc_param struct that indicates if it was user_provided and acts upon it in the rebuild use case